service/raft/raft_group0.cc

```diff
@@ -374,33 +374,32 @@ future<> raft_group0::become_voter() {
 }
 
 future<> raft_group0::leave_group0() {
+    assert(this_shard_id() == 0);
+
     if (!_raft_gr.is_enabled()) {
+        rslog.info("leave_group0: local RAFT feature disabled, skipping.");
         co_return;
     }
-    assert(this_shard_id() == 0);
-    raft::server_id remove_addr;
-    auto my_id = raft::server_id{co_await db::system_keyspace::get_raft_server_id()};
-    if (my_id) {
-        remove_addr = my_id;
-    } else {
-        // Nothing to do
+
+    if (!joined_group0()) {
+        // TODO: unimplemented upgrade case.
         co_return;
     }
-    auto pause_shutdown = _shutdown_gate.hold();
-    if (std::holds_alternative<raft::group_id>(_group0)) {
-        co_return co_await _raft_gr.group0().modify_config({}, {remove_addr}, &_abort_source);
-    }
-    auto g0_info = co_await discover_group0(my_id, raft::server_info{} /* XXX: apparently passing empty info works */);
```

Comment:
The reason discover needs to run when removing self is recovering from a failed boot. Of course you could ban it, then, if a node boot fails, you'd have to go to another node to remove it. But then you won't find it in the peers table on that other node, since its boot failed (it may have only succeeded joining group0, but not adding itself to the peers). 

Of course we would not have to have this conversation if there was a test for a failed join_group0< which we still don't have :/

Reply:
To run `decommission`, you *must* finish bootstrap first.
Therefore, if a node fails boot, you won't be able to run decommission without restarting and finishing the boot first - at which point the node will become a member of group 0.

Reply:
BTW. that's why I think joining group 0 should be a step in the topology change saga. Then recovering from failed boots will be automatically handled by the topology change coordinator.

Reply:
Discovery algorithm tells you whether you should create group 0 or whether you should join existing group 0.
If we should create group 0, then we can create group 0, start the coordinator, and pass control to the coordinator.
If we should join, we can pass control to the coordinator immediately.

Reply:
And this discover algorithm should handle restarts anyway whether it is a first node or not the first without any external coordinator.

Reply:
Yes, it handles restarts and doesn't depend on any coordinator.

The problem is with the next step - joining group 0, which changes the group 0 configuration. If the node crashes at this point and isn't restarted, group 0 config contains garbage. The coordinator could clean this garbage.

Reply:
A coordinator does not have magic forecast capability. Removing the node will be an administrative decision.

Reply:
You want to make the situation worse than it was.

Today if bootstrap fails and admin gives up on the node, the cluster will eventually clean up after the node, removing information from gossiper application state and so on. It takes about a minute after the node becomes unresponsive I think (2*ring_delay or sth).

Now we want to change this situation - the admin can no longer simply give up on the node, now they have to run additional clean up steps?

If we follow @tgrabiec design (I know you don't like it - but let's suppose we do), the coordinator drives the entire saga. If the coordinator notices at some point that the added node becomes unresponsive, the coordinator enters one of the abort steps of the saga (which step depends on how far in the saga we went). We can simply extend this design with joining group 0 step at the beginning, and have a corresponding abort step as well. So the coordinator tries to join the added node to group 0, but if it keeps failing, eventually it cleans up and finishes.

Reply:
> You want to make the situation worse than it was.
> 
> Today if bootstrap fails and admin gives up on the node, the cluster will eventually clean up after the node, removing information from gossiper application state and so on. It takes about a minute after the node becomes unresponsive I think (2*ring_delay or sth).
> 
And that may cause inconsistencies if the node restart after the delay. Do not remember details but remember discussing it with Asias.

Reply:
Inconsistencies which we will fix thanks to centralized coordinator.

Reply:
Not necessary. You cannot say something will be fixed only because of a coordinator.  For instance current inconsistency can also be fixed by fencing, no coordinator needed.  Anyway it is orthogonal who decides to joint to group0 and who decide to remove. A cluster may have a heuristic to remove nodes automatically following whatever criteria. The heuristic may or may not be executed by the topology change coordinator. The important part is to be able to fence the removed node and this is not done by the coordinator, but by the entire cluster.

Reply:
I thought the entire point of the coordinator is to put all decisions about topology in a single totally ordered sequence. And every decision in this sequence knows about all decisions earlier in the sequence. Then you cannot make decisions that are incompatible with other decisions, because there is no concurrency in the decision making process.

If we allow concurrency between deciding who to join and who to remove, we're basically making our lives harder. What for?

Reply:
Decisions about topology can be made by multiple admins simultaneously and that purpose to move topology to raft is to make them linearizable.  The coordinator has nothing to do with linearizablility.

Reply:
Now you changed the meaning of "decisions" to argue your point.

I'm talking about "decisions" as you used it here:
> who decides to joint to group0 and who decide to remove

I.e. there are two meanings of "decision"
1. Decision to execute a topology change operation, which consists of a sequence of smaller steps ("saga"). Can be made by admins concurrently
2. Decision to execute a single step of topology change saga.

We were talking about the second meaning. The coordinator's purpose is to put these decisions in a single sequence and have each decision be compatible with all earlier decisions.

Reply:
> Now you changed the meaning of "decisions" to argue your point.

No, I did not.

> 
> I'm talking about "decisions" as you used it here:
> 
> > who decides to joint to group0 and who decide to remove
> 
> I.e. there are two meanings of "decision"
> 
>     1. Decision to execute a topology change operation, which consists of a sequence of smaller steps ("saga"). Can be made by admins concurrently

This is the only meaning I ever used. Remember that we discussed admin decision to remove failed node? This is a decision to  execute topology change operation. It does not matter if admin decides it, some kind of script or some internal logic in Scylla. 

> 
>     2. Decision to execute a single step of topology change saga. 
> 
> 
> We were talking about the second meaning. The coordinator's purpose is to put these decisions in a single sequence and have each decision be compatible with all earlier decision 

The coordinator purpose is to boostrap joining nodes (mostly means re-distributing data in the cluster). The order is created by raft not by the coordinator. The coordinator may some day allow bootstrapping multiple nodes simultaneously (frankly I think this is the main reason it was introduced at all)  since then it may re-distribute data more efficiently, but then, by definition, joining group0 for a particular node is a separate operation from bootstraping the node.

Reply:
> Remember that we discussed admin decision to remove failed node? This is a decision to execute topology change operation.

> (...) but then, by definition, joining group0 for a particular node is a separate operation from bootstraping the node.

Let us distinguish two things:
1. "atomic" topology change steps
2. non-atomic multi-step topology change operations

Bootstrapping is a non-atomic multi-step topology change operation.
It consists of the following "atomic" steps: join group 0, update token ring, stream data, update token ring again, become voter.

The admin doesn't request atomic steps, the admin requests topology change operations (which are multi-step).

Any node can handle a request from the administrator and start a (non-atomic multi-step) topology change operation. The coordinator's purpose is to implement those operations by performing a sequence of "atomic" steps. The handling node starts an operation by inserting it into the coordinator's queue.

My argument is as follows:
1. Joining group 0 is one of those "atomic" steps.
2. Bootstrap is a non-atomic multi-step operation so by definition it consists of a sequence of atomic steps.
3. The first step in that sequence is joining group 0.
4. A non-atomic multi-step operation succeeds when it performs each atomic step in the sequence.
5. If one of the steps fails or it's impossible to proceed with it due to some unavailability, the non-atomic multi-step operation fails.
6. If a non-atomic multi-step operation fails, we should clean up by rolling back each atomic step that was performed up to this point.
7. Therefore, if bootstrap fails, we should rollback the join group 0 step.
8. The coordinator is responsible for performing the atomic steps, hence it is responsible for rolling back the group 0 step as well.

Do you agree with these points? If you disagree - which points do you disagree with? Or maybe you disagree with the entire model, e.g. this is not what you think the coordinator should do or how we should look at topology change operations?

Reply:
> Do you agree with these points? If you disagree - which points do you disagree with? Or maybe you disagree with the entire model, e.g. this is not what you think the coordinator should do or how we should look at topology change operations?

Joining group0 is not part of the bootstrap.


> > Remember that we discussed admin decision to remove failed node? This is a decision to execute topology change operation.
> 
> > (...) but then, by definition, joining group0 for a particular node is a separate operation from bootstraping the node.
> 
> Let us distinguish two things:
> 
>     1. "atomic" topology change steps
> 
>     2. non-atomic multi-step topology change operations
> 
> 
> Bootstrapping is a non-atomic multi-step topology change operation. It consists of the following "atomic" steps: join group 0, update token ring, stream data, update token ring again, become voter.

Joining group0 is not part of the bootstrap. It is a cluster topology change, but actual boostrap my happen later. May be after some more nodes join, may be when tablets logic decides to move some ranges to the node. You makes things more complicated by mixing different concepts.

> 
> The admin doesn't request atomic steps, the admin requests topology change operations (which are multi-step).

And removing a partially joined node may be a multi step operation. The important point is the decision to get rid of the node. This is an administrative decision even if done by automatic logic. In fact an admin by give such command in the middle of a successful boostrap for some reason. Or an admin may want to remove the node without waiting for the timeout after which Scylla decided to do it because it wants to start another boostrap and we currently allow only one.
 
> 
> Any node can handle a request from the administrator and start a (non-atomic multi-step) topology change operation. The coordinator's purpose is to implement those operations by performing a sequence of "atomic" steps. The handling node starts an operation by inserting it into the coordinator's queue.
> 
> My argument is as follows:
> 
>     1. Joining group 0 is one of those "atomic" steps.
> 
>     2. Bootstrap is a non-atomic multi-step operation so by definition it consists of a sequence of atomic steps.
> 
>     3. The first step in that sequence is joining group 0.
> 
>     4. A non-atomic multi-step operation succeeds when it performs each atomic step in the sequence.
> 
>     5. If one of the steps fails or it's impossible to proceed with it due to some unavailability, the non-atomic multi-step operation fails.
> 
>     6. If a non-atomic multi-step operation fails, we should clean up by rolling back each atomic step that was performed up to this point.
> 
>     7. Therefore, if bootstrap fails, we should rollback the join group 0 step.
> 
>     8. The coordinator is responsible for performing the atomic steps, hence it is responsible for rolling back the group 0 step as well.
> 
> 
> Do you agree with these points? If you disagree - which points do you disagree with? Or maybe you disagree with the entire model, e.g. this is not what you think the coordinator should do or how we should look at topology change operations?

I disagree with this very simplistic model that does not take into account neither tablets nor simultaneous bootstrap. I also disagree with naive assumption that to abort boostrap you need to rollback step by step. No, you just decide to remove partially bootstrapped node and do it by whatever means needed, not rolling back steps that may be irrelevant (why would you stream data back from a node you decided to not bootstrap for instance?).

Reply:
> To run `decommission`, you _must_ finish bootstrap first. Therefore, if a node fails boot, you won't be able to run decommission without restarting and finishing the boot first - at which point the node will become a member of group 0.

I was considering the case when this node's bootstrap is stuck for whatever reason. You can't decommission it on some other nodes because other nodes don't have it in peers yet. But api::storage_service::decommission is available as a local rest endpoint before you've fully joined. 

In other words, this all had to do with deciding which configuration is the primary source of truth - peers or raft config. I assumed they can be out of sync, at least theoretically, so if ther eis a request to decommission a node which is not part of group0, we must make a fair attempt to remove it. 

All of the above was discussed during the review. It's fine to remove it - but until we have a test case, we could be going back and forth adding and removing it forever, without a test this is dead code.

Reply:
> Joining group0 is not part of the bootstrap.

Today it is. A node cannot join group 0 in any other way than through bootstrapping.
Of course we can modify the set of available operations and split "bootstrap" from join group 0 in the future.

> May be after some more nodes join, may be when tablets logic decides to move some ranges to the node. You makes things more complicated by mixing different concepts.

You're describing a possible future where after joining group 0 a node is already a "normal member" of the cluster.

> And removing a partially joined node may be a multi step operation. The important point is the decision to get rid of the node. This is an administrative decision even if done by automatic logic. In fact an admin by give such command in the middle of a successful boostrap for some reason. Or an admin may want to remove the node without waiting for the timeout after which Scylla decided to do it because it wants to start another boostrap and we currently allow only one.

What you're describing is cancelling bootstrap. We can handle cancelling bootstrap in exactly the same way as we handle a partially failed bootstrap - by rolling back saga steps. Just the trigger is different, one is automatic, the other is manual by admin.

> I disagree with this very simplistic model that does not take into account neither tablets nor simultaneous bootstrap.

This model was created with allowing us to introduce tablet-manipulating steps and simultaneous bootstrap steps into the sagas in mind - in the future. With the saga model in mind, we only need to decide what the steps and operations are and which operation uses which steps.

> I also disagree with naive assumption that to abort boostrap you need to rollback step by step. No, you just decide to remove partially bootstrapped node and do it by whatever means needed, not rolling back steps that may be irrelevant (why would you stream data back from a node you decided to not bootstrap for instance?).

Who said rolling back streaming means streaming data back? I didn't.
Have you ever read Tomek's designs or specifications? He includes the aborting steps in those designs.

```python3
def abort_after_streaming(tx: TransactionId, coid: CoordinatorId, t: Timestamp):
    set_stage(tx, ReplicationStage.write_both_read_old, t)
    return 'abort_streaming'


def abort_streaming(tx: TransactionId, coid: CoordinatorId, t: Timestamp):
    stop_streaming(tx)
    return 'abort_advertise_ring'


def abort_advertise_ring(tx: TransactionId, coid: CoordinatorId, t: Timestamp):
    replicate_token_metadata(as_mutation(get_old_ring(local_ring()), t))
    return 'abort_update_raft_group'


def abort_update_raft_group(tx: TransactionId, coid: CoordinatorId, t: Timestamp):
    for n in local_ring().joining_members():
        this_node.raft_gr0.remove_node(n)
    return 'unlock'

# Each step should be idempotent and have no effect on end result in case of preemption and failover.
topology_change_state_machine = {
    'update_raft_group': step_update_raft_group,
    'advertise_ring': step_advertise_ring,
    'streaming': step_streaming,
    'after_streaming': step_after_streaming,
    'use_only_new': step_use_only_new,
    'mark_dead': step_mark_dead,
    'cleanup': step_cleanup,
    'only_new_ring': step_only_new_ring,
    'unlock': step_unlock,
    'done': step_done,

    # Abort steps
    'abort_after_streaming': abort_after_streaming,
    'abort_streaming': abort_streaming,
    'abort_advertise_ring': abort_advertise_ring,
    'abort_update_raft_group': abort_update_raft_group,
}

 abort_steps = {
        'after_streaming': 'abort_after_streaming',
        'streaming': 'abort_streaming',
        'advertise_ring': 'abort_advertise_ring',
        'update_raft_group': 'abort_update_raft_group',
    }
```

If `after_streaming` fails, we go into `abort_after_streaming`.
If `streaming` fails, we go into `abort_streaming`. Observe that `abort_streaming` does not stream back, it simply stops any existing streaming operations.
And so on.

from https://github.com/tgrabiec/scylla/blob/safe-topology-changes/docs/safe-ring-changes.py

Reply:
> > Joining group0 is not part of the bootstrap.
> 
> Today it is. A node cannot join group 0 in any other way than through bootstrapping. Of course we can modify the set of available operations and split "bootstrap" from join group 0 in the future.

Today _you_ look at it as part of the boostrap. I gave you another perspective. It is not about the code, but how you think about it. Yes, today what I called boostrapping starts immediately after the joining because this is the only mode of operation supported, but it does not make one be part of the other.

> 
> > May be after some more nodes join, may be when tablets logic decides to move some ranges to the node. You makes things more complicated by mixing different concepts.
> 
> You're describing a possible future where after joining group 0 a node is already a "normal member" of the cluster.
> 

OK, this is the direction.

> > And removing a partially joined node may be a multi step operation. The important point is the decision to get rid of the node. This is an administrative decision even if done by automatic logic. In fact an admin by give such command in the middle of a successful boostrap for some reason. Or an admin may want to remove the node without waiting for the timeout after which Scylla decided to do it because it wants to start another boostrap and we currently allow only one.
> 
> What you're describing is cancelling bootstrap. We can handle cancelling bootstrap in exactly the same way as we handle a partially failed bootstrap - by rolling back saga steps. Just the trigger is different, one is automatic, the other is manual by admin.

It does not matter how you call it. The point is that it is an administrative decision.  RAFT makes sure the end result is consistent, not the coordinator. The coordinator is more for a liveness purposes.


> 
> > I disagree with this very simplistic model that does not take into account neither tablets nor simultaneous bootstrap.
> 
> This model was created with allowing us to introduce tablet-manipulating steps and simultaneous bootstrap steps into the sagas in mind - in the future. With the saga model in mind, we only need to decide what the steps and operations are and which operation uses which steps.
> 
> > I also disagree with naive assumption that to abort boostrap you need to rollback step by step. No, you just decide to remove partially bootstrapped node and do it by whatever means needed, not rolling back steps that may be irrelevant (why would you stream data back from a node you decided to not bootstrap for instance?).
> 
> Who said rolling back streaming means streaming data back? I didn't. Have you ever read Tomek's designs or specifications? He includes the aborting steps in those designs.

Python code is neither design nor specification, so no I did not read it and do not mean to. My point is that aborting boostrap does not necessary done via rolling back steps.  Requiring it in a design just puts unnecessary burden on the implementation.

Reply:
> Python code is neither design nor specification, 

Do you think English is better? Because English, as natural human languages all do, suffers from ambiguities. Programming languages are better that way because you need to be precise (otherwise your compiler / type checker / linter / whatever) will reject it.

Would you look at the spec if I rewrote it, say, in TLA+? I could do it. It won't look very different from this Python code though.

> so no I did not read it and do not mean to.

Come on.

> My point is that aborting boostrap does not necessary done via rolling back steps. 

That's just miscommunication between us. Let's not call it "rolling back" then because you think rolling back must mean performing the step "in reverse", e.g. rolling back streaming must mean streaming data back.

Let's call it "aborting steps".

And some aborting steps ARE necessary - e.g. if you added pending tokens to the token ring and fail bootstrap, you must remove them from the token ring. That's an aborting step.

> Requiring it in a design just puts unnecessary burden on the implementation.

You'll have to do the design at some point, I think it's good we do the design earlier so we can discuss it before we write code.